### ISSUE TITLE
GStringExpressionWithinString false positive for an escaped $ character within a GString

### ISSUE THREAD

**Issue by erdi** [CodeNarc/CodeNarc#188]
I'm getting a GStringExpressionWithinString violation for a string like:
`"$valueToBeReplaced \$valueNotToBeReplaced"` which clearly is a GString with an escaped `$` character in it.

**Comment by chrismair** [comment]
I added this to the tests for that rule:

```
        final SOURCE = '''
            def valueToBeReplaced = '123'
            ...
            def str8 = "$valueToBeReplaced \$valueNotToBeReplaced"
        '''
        assertNoViolations(SOURCE)
```
and it works as expected -- no violation.

That rule should only apply to single-quote strings. Were you actually using single-quotes? If so, then I think the violation would be legitimate.

**Comment by erdi** [comment]
Sorry for not being clear when submitting the issue. I think if you change your test case  to:
```
final SOURCE = '''
    def str8 = "\$valueNotToBeReplaced"
'''
```
You will get a violation.

I'm getting a violation for the following code: https://github.com/energizedwork/asl2-gradle-plugin/blob/8e03d30115caf9eb5e502f0ae3df45946bd14cca/src/test/groovy/com/energizedwork/gradle/IdeaCopyrightHeadersConfigSpec.groovy#L46

To reproduce you can simply clone that repo, remove suppression for that rule from `gradle/codenarc/rulesets.groovy` and run `./gradlew check`.

**Comment by chrismair** [comment]
Okay thanks. That helped. I cloned your project and reproduced the violation, and then was able to reproduce within a CodeNarc rule test.

So, the behavior is from some interesting AST parsing in Groovy.

This whole thing is a GString, as expected:
```
"""
        task $testTaskName {
            doLast {
                println "Has idea plugin: \${project.plugins.hasPlugin(IdeaPlugin)}"
            }
        }
    """
```
But that GString is parsed into multiple component _strings_, including this:

`                    task $`

and this:

```
{
            doLast {
                println "Has idea plugin: \${project.plugins.hasPlugin(IdeaPlugin)}"
            }
        }
```

and each of those is a _regular_ String. Given the relatively simplistic parsing done by the CodeNarc rule, that explains the violation. But that rule needs to get a bit more sophisticated.

**Comment by chrismair** [comment]
Another issue is that escaping a $ within a regular string has no effect. Both of these statements print the same thing:
```
println 'An example GString is \${count}'
println 'An example GString is ${count}'
```
and that "\" is omitted in the value of the string returned from the AST parser.

**Comment by chrismair** [comment]
Fixed for CodeNarc 0.27. The rule now skips strings contained within GStrings.
